Source/KSCrash/Source/KSCrash/Recording/BSG_KSMachHeaders.m

```diff
@@ -0,0 +1,190 @@
+}
+
+uintptr_t bsg_ksdlfirstCmdAfterHeader(const struct mach_header *const header);
+
+/**
+ * Populate a Mach binary image info structure
+ *
+ * @param header The Mach binary image header
+ *
+ * @param slide The VM offset of the binary image
+ *
+ * @param info A reference to the structure
+ *
+ * @returns a boolean indicating success
+ */
+bool populate_info(const struct mach_header *header, BSG_Mach_Binary_Image_Info *info) {
+    
+    // Early exit conditions; this is not a valid/useful binary image
+    // 1. We can't find a sensible Mach command
+    uintptr_t cmdPtr = bsg_ksdlfirstCmdAfterHeader(header);
+    if (cmdPtr == 0) {
+        return false;
+    }
+
+    // 2. The image doesn't have a name.  Note: running with a debugger attached causes this condition to match.
+    Dl_info DlInfo = (const Dl_info) { 0 };
+    dladdr(header, &DlInfo);
+    const char *image_name = DlInfo.dli_fname;
+    if (!image_name) {
+        return false;
+    }
+    
+    // Look for the TEXT segment to get the image size.
+    // Also look for a UUID command.
+    uint64_t imageSize = 0;
+    uint64_t imageVmAddr = 0;
+    uint8_t *uuid = NULL;
+
+    for (uint32_t iCmd = 0; iCmd < header->ncmds; iCmd++) {
+        struct load_command *loadCmd = (struct load_command *)cmdPtr;
+        switch (loadCmd->cmd) {
+        case LC_SEGMENT: {
+            struct segment_command *segCmd = (struct segment_command *)cmdPtr;
+            if (strcmp(segCmd->segname, SEG_TEXT) == 0) {
+                imageSize = segCmd->vmsize;
+                imageVmAddr = segCmd->vmaddr;
+            }
+            break;
+        }
+        case LC_SEGMENT_64: {
+            struct segment_command_64 *segCmd =
+                (struct segment_command_64 *)cmdPtr;
+            if (strcmp(segCmd->segname, SEG_TEXT) == 0) {
+                imageSize = segCmd->vmsize;
+                imageVmAddr = segCmd->vmaddr;
+            }
+            break;
+        }
+        case LC_UUID: {
+            struct uuid_command *uuidCmd = (struct uuid_command *)cmdPtr;
+            uuid = uuidCmd->uuid;
+            break;
+        }
+        }
+        cmdPtr += loadCmd->cmdsize;
+    }
+    
+    // Save these values
+    info->mh = header;
+    info->cpusubtype = header->cpusubtype;
+    info->cputype = header->cputype;
+    info->imageSize = imageSize;
+    info->imageVmAddr = imageVmAddr;
+    info->uuid = uuid;
+    info->name = image_name;
+    
+    return true;
+}
+
+/**
+ * A callback invoked when dyld loads binary images.  It stores enough relevant info about the
+ * image to populate a crash report later.
+ *
+ * @param header A mach_header structure
+ *
+ * @param slide The VM offset of the binary image
+ */
+void bsg_mach_binary_image_added(const struct mach_header *header, intptr_t slide)
+{
+    BSG_Mach_Binary_Image_Info info = { 0 };
+    if (populate_info(header, &info)) {
+        [bsg_mach_binary_images_info addObject:[NSValue valueWithBytes:&info
```

Comment:
Using an Objective-C object makes me a bit nervous - as if `dyld` continues to execute this callback during a signal crash, then it could hang the process.

Would it be possible to initialise a C array to start with, then allocate more memory as required?